### ISSUE TITLE
Issue #16781: `NoWhitespaceBefore` blind spot resulting in high failure rate

### ISSUE THREAD

**Issue by pankratz76** [checkstyle/checkstyle#16820]
I have read check documentation: https://checkstyle.org/checks/xxxxxx/nameofaffectedcheck.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac YOUR_FILE.java
#[[MAKE SURE THERE IS SUCCESSFUL COMPILATION]]

/var/tmp $ cat config.xml
#[[PLACE YOUR OUTPUT HERE]]

/var/tmp $ cat YOUR_FILE.java
/*
NoWhitespaceBefore
allowLineBreaks = (default)false
tokens = (default)COMMA, SEMI, POST_INC, POST_DEC, ELLIPSIS, LABELED_STAT


*/

package com.puppycrawl.tools.checkstyle.checks.whitespace.nowhitespacebefore;

/**
 * Class for testing whitespace issues.
 * violation missing author tag
 **/
class InputNoWhitespaceBeforeDefaultNextGeneration
{
    public void test() {
        // Valid cases
        " ".equals("");
        "".equals(" ");
        "".equals("");

        // Violations
        "".equals("");
        "".equals(""  + ""); // violation
        "".equals("" +  ""); // violation
         System.out.println(""); // violation
        System .out.println(""); // violation
        System. out.println(""); // violation
        System.out .println(""); // violation
        System.out. println(""); // violation
        System.out. println(""); // violation
        System.out.println (""); // violation
        "" .equals(""); // violation
        "". equals(""); // violation
        "".equals (""); // violation
        "".equals( ""); // violation
        "".equals("" ); // violation
        "".equals("") ; // violation
    }

    /** Test variable assignment violations */
    public void testSpaceViolationVarAssignment() {
        // Valid
        boolean eq1 = " ".equals("");
        boolean eq2 = "".equals("");

        // Violations
        eq2 = "".equals(""); // violation
        eq2  = "".equals(""); // violation
        eq2 =  "".equals(""); // violation
    }

    /** Test variable declaration violations */
    public void testSpaceViolationVarDeclaration() {
        // Valid
        boolean e = "".equals("");
        boolean e3 = "".equals("");
        e3 = "".equals("");

        // Violations
        boolean e4 = "".equals(""); // violation
        boolean e1  = "".equals(""); // violation
        boolean  e2 = "".equals(""); // violation
        e3 = "".equals(""); // violation
        e3  = "".equals(""); // violation
        e3 =  "".equals(""); // violation
    }

    /** Test array access violations */
    public void testArrayAccess() {
        int[] arr = new int[10];

        // Valid
        int a = arr[0];

        // Violations
        int x = arr [0]; // violation
        int y = arr[ 0]; // violation
        int z = arr[0]; // violation
    }

    /** Test generics violations */
    public void testGenerics() {
        java.util.List<String> list = new java.util.ArrayList<>();

        // Valid
        list.add("test");

        // Violations
        list .add("test"); // violation
        list. add("test"); // violation
    }

    /** Test lambda violations */
    public void testLambda() {
        // Valid
        Runnable r = () -> System.out.println();

        // Violation
        Runnable r2 = () -> System.out .println(); // violation
    }

    /** Test method reference violations */
    public void testMethodReference() {
        // Valid
        java.util.function.Function<String, String> f1 = String::valueOf;

        // Violation
        java.util.function.Function<String, String> f2 = String ::valueOf; // violation
    }

    /** Test nested calls violations */
    public void testNestedCalls() {
        String s = "hello";

        // Valid
        s.substring(1).trim();
        s.substring(1).trim();

        // Violations
        s.substring(1). trim(); // violation
        s.substring(1 ).trim(); // violation
    }

    /** Test multiple dots violations */
    public void testMultipleDots() {
        String s = "hello";

        // Valid
        s.substring(1).substring(1).substring(1);

        // Violations
        s.substring(1).substring(1). substring(1); // violation
        s.substring(1) .substring(1).substring(1); // violation
    }

    /** Test with other operators */
    public void testWithOtherOperators() {
        // Valid
        String s1 = "a" + "b".toString();
        int x = 1 + 2 * 3;

        // Violation
        String s2 = "a" + "b". toString(); // violation
    }

    /** Test in control structures */
    public void testInControlStructures() {
        // Valid
        if ("test".equals("test")) {
            System.out.println();
        }

        // Violations
        if ("test". equals("test")) { // violation
            System.out.println();
        }

        while (true) {
            break ;
        }
    }

    /** Test in try-catch */
    public void testInTryCatch() {
        // Valid
        try {
            // do something
        } catch (Exception e) {
            // handle
        }

        // Violation
        try {
            // do something
        } catch (Exception e ) { // violation
            // handle
        }
    }

    /** Test in annotations */
    public void testInAnnotations() {
        // Valid
        @SuppressWarnings("unchecked")
        Object o1 = new Object();

        // Violation
        @SuppressWarnings ("unchecked") // violation
        Object o2 = new Object();
    }

    /** Test in type cast */
    public void testInTypeCast() {
        Object o = "test";

        // Valid
        String s1 = (String) o;

        // Violation
        String s2 = (String ) o; // violation
    }

    /** Test in switch */
    public void testInSwitch() {
        // Valid
        switch (1) {
            case 1:
                break;
        }

        // Violation
        switch (1 ) { // violation
            case 1:
                break;
        }
    }

    /** Test in synchronized */
    public void testInSynchronized() {
        // Valid
        synchronized (this) {
            // do something
        }

        // Violation
        synchronized (this ) { // violation
            // do something
        }
    }

    /** Test in assert */
    public void testInAssert() {
        // Valid
        assert true : "message";

        // Violation
        assert true : "message" ; // violation
    }

    /** Test in return */
    public void testInReturn() {
        // Valid (no return value)
        return;
    }

    public void testInReturn2() {
        // Violation
        return ; // violation
    }

    /** Test in throw */
    public void testInThrow() throws Exception {
        // Valid
        throw new Exception();
    }

    public void testInThrow2() throws Exception {
        // Violation
        throw new Exception() ; // violation
    }

    /** Test in array initializer */
    public void testInArrayInitializer() {
        // Valid
        int[] arr1 = new int[] {1, 2, 3};

        // Violation
        int[] arr2 = new int[] {1, 2, 3 } ; // violation
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-X.XX-all.jar -c config.xml YOUR_FILE.java
#[[PLACE YOUR OUTPUT HERE]]
```

For Windows users, please use `type` instead of `cat` and run

```cmd
set RUN_LOCALE="-Duser.language=en -Duser.country=US"
java %RUN_LOCALE% -jar checkstyle-X.XX-all.jar -c config.xml YOUR_FILE.java
```

in place of the last 2 commands above.

---

**Describe what you expect in detail.**

---

Still not clear ???
see example - https://checkstyle.org/report_issue.html#How_to_report_a_bug.3F

ATTENTION: FAILURE TO FOLLOW THE ABOVE TEMPLATE WILL RESULT IN THE ISSUE BEING CLOSED.

**Comment by pankratz76** [comment]
```
/*
NoWhitespaceBefore
allowLineBreaks = (default)false
tokens = (default)COMMA, SEMI, POST_INC, POST_DEC, ELLIPSIS, LABELED_STAT


*/

package com.puppycrawl.tools.checkstyle.checks.whitespace.nowhitespacebefore;

/**
 * Class for testing whitespace issues.
 * violation missing author tag
 **/
class InputNoWhitespaceBeforeDefaultNextGeneration
{
    public void test() {
        // Valid cases
        " ".equals("");
        "".equals(" ");
        "".equals("");

        // Violations
        "".equals("");
        "".equals(""  + ""); // violation
        "".equals("" +  ""); // violation
         System.out.println(""); // violation
        System .out.println(""); // violation
        System. out.println(""); // violation
        System.out .println(""); // violation
        System.out. println(""); // violation
        System.out. println(""); // violation
        System.out.println (""); // violation
        "" .equals(""); // violation
        "". equals(""); // violation
        "".equals (""); // violation
        "".equals( ""); // violation
        "".equals("" ); // violation
        "".equals("") ; // violation
    }

    /** Test variable assignment violations */
    public void testSpaceViolationVarAssignment() {
        // Valid
        boolean eq1 = " ".equals("");
        boolean eq2 = "".equals("");

        // Violations
        eq2 = "".equals(""); // violation
        eq2  = "".equals(""); // violation
        eq2 =  "".equals(""); // violation
    }

    /** Test variable declaration violations */
    public void testSpaceViolationVarDeclaration() {
        // Valid
        boolean e = "".equals("");
        boolean e3 = "".equals("");
        e3 = "".equals("");

        // Violations
        boolean e4 = "".equals(""); // violation
        boolean e1  = "".equals(""); // violation
        boolean  e2 = "".equals(""); // violation
        e3 = "".equals(""); // violation
        e3  = "".equals(""); // violation
        e3 =  "".equals(""); // violation
    }

    /** Test array access violations */
    public void testArrayAccess() {
        int[] arr = new int[10];

        // Valid
        int a = arr[0];

        // Violations
        int x = arr [0]; // violation
        int y = arr[ 0]; // violation
        int z = arr[0]; // violation
    }

    /** Test generics violations */
    public void testGenerics() {
        java.util.List<String> list = new java.util.ArrayList<>();

        // Valid
        list.add("test");

        // Violations
        list .add("test"); // violation
        list. add("test"); // violation
    }

    /** Test lambda violations */
    public void testLambda() {
        // Valid
        Runnable r = () -> System.out.println();

        // Violation
        Runnable r2 = () -> System.out .println(); // violation
    }

    /** Test method reference violations */
    public void testMethodReference() {
        // Valid
        java.util.function.Function<String, String> f1 = String::valueOf;

        // Violation
        java.util.function.Function<String, String> f2 = String ::valueOf; // violation
    }

    /** Test nested calls violations */
    public void testNestedCalls() {
        String s = "hello";

        // Valid
        s.substring(1).trim();
        s.substring(1).trim();

        // Violations
        s.substring(1). trim(); // violation
        s.substring(1 ).trim(); // violation
    }

    /** Test multiple dots violations */
    public void testMultipleDots() {
        String s = "hello";

        // Valid
        s.substring(1).substring(1).substring(1);

        // Violations
        s.substring(1).substring(1). substring(1); // violation
        s.substring(1) .substring(1).substring(1); // violation
    }

    /** Test with other operators */
    public void testWithOtherOperators() {
        // Valid
        String s1 = "a" + "b".toString();
        int x = 1 + 2 * 3;

        // Violation
        String s2 = "a" + "b". toString(); // violation
    }

    /** Test in control structures */
    public void testInControlStructures() {
        // Valid
        if ("test".equals("test")) {
            System.out.println();
        }

        // Violations
        if ("test". equals("test")) { // violation
            System.out.println();
        }

        while (true) {
            break ;
        }
    }

    /** Test in try-catch */
    public void testInTryCatch() {
        // Valid
        try {
            // do something
        } catch (Exception e) {
            // handle
        }

        // Violation
        try {
            // do something
        } catch (Exception e ) { // violation
            // handle
        }
    }

    /** Test in annotations */
    public void testInAnnotations() {
        // Valid
        @SuppressWarnings("unchecked")
        Object o1 = new Object();

        // Violation
        @SuppressWarnings ("unchecked") // violation
        Object o2 = new Object();
    }

    /** Test in type cast */
    public void testInTypeCast() {
        Object o = "test";

        // Valid
        String s1 = (String) o;

        // Violation
        String s2 = (String ) o; // violation
    }

    /** Test in switch */
    public void testInSwitch() {
        // Valid
        switch (1) {
            case 1:
                break;
        }

        // Violation
        switch (1 ) { // violation
            case 1:
                break;
        }
    }

    /** Test in synchronized */
    public void testInSynchronized() {
        // Valid
        synchronized (this) {
            // do something
        }

        // Violation
        synchronized (this ) { // violation
            // do something
        }
    }

    /** Test in assert */
    public void testInAssert() {
        // Valid
        assert true : "message";

        // Violation
        assert true : "message" ; // violation
    }

    /** Test in return */
    public void testInReturn() {
        // Valid (no return value)
        return;
    }

    public void testInReturn2() {
        // Violation
        return ; // violation
    }

    /** Test in throw */
    public void testInThrow() throws Exception {
        // Valid
        throw new Exception();
    }

    public void testInThrow2() throws Exception {
        // Violation
        throw new Exception() ; // violation
    }

    /** Test in array initializer */
    public void testInArrayInitializer() {
        // Valid
        int[] arr1 = new int[] {1, 2, 3};

        // Violation
        int[] arr2 = new int[] {1, 2, 3 } ; // violation
    }
}
```

**Comment by pankratz76** [comment]
```
    public void test() {
        // Valid cases
        " ".equals("");
        "".equals(" ");
        "".equals("");

        // Violations
        "".equals("");
        "".equals(""  + ""); // violation
        "".equals("" +  ""); // violation
         System.out.println(""); // violation
        System .out.println(""); // violation
        System. out.println(""); // violation
        System.out .println(""); // violation
        System.out. println(""); // violation
        System.out. println(""); // violation
        System.out.println (""); // violation
        "" .equals(""); // violation
        "". equals(""); // violation
        "".equals (""); // violation
        "".equals( ""); // violation
        "".equals("" ); // violation
        "".equals("") ; // violation
    }
```

**Comment by pankratz76** [comment]
I'm not sure what the issue is here—because to me, it seems suspicious that the checker isn't finding these violations when executing this locally, just like in [#16810](https://github.com/checkstyle/checkstyle/pull/16810).

You can easily check out the branch from the PR yourself and run it locally. Alternatively, just copy and paste the `test` method into the `InputNoWhitespaceBeforeDefault.java` file and run the test to see for yourself.

At this point, if there's still hesitation to acknowledge this, I'm not sure whether it's a communication issue or something else—if you know what I mean. I'm genuinely trying to help and contribute as best as I can.

**Comment by nrmancuso** [comment]
@pankratz76 

**Please keep your tone respectful.** We are all volunteers here in this community, and abrasive or sarcastic comments are generally unproductive; they will not help to move this forward.

The issue template exists to save everyone's time (including yours). Whatever urgency you feel around this doesn’t exempt you from following it. If you're serious about resolving this, I ask that you follow the process.

Once you've done that, I would be happy to take a closer look and help to move things forward.

**Comment by nrmancuso** [comment]
I am closing this since it is a duplicate of https://github.com/checkstyle/checkstyle/issues/16781

**Comment by pankratz76** [comment]
> If you're serious about resolving this, I ask that you follow the process.

Same for you - Please provide the template by checking out my example or show me once so I can learn from it.

I've thought deeply about this and believe we can fix the 95% blind spot in this problematic code, including tests. It's shocking how far it is from its purpose, but we improve through failure.

Why is template process better than a working PR? Checking a branch and running tests seems faster than template processing. A failing test is all we need (#16823 proves this).

Sorry for frustration - I felt bullied by process over solutions. The tests (the "judge") prove my case. Fixing these tests will fix implementation.

The 95% failure rate suggests test issues - 5% might be acceptable, but not 95% in a high quality project like this beauty.

We want the same goal, even if we disagree on approach.

Pick yourself just one line from the example and it will fail instantly:

Please don't act like you don't have opportunity to recreate it yourself in seconds. I have tried it like crazy to provide you everything needed.

At rewrite we work exact like this, without any issue:
https://github.com/openrewrite/rewrite-static-analysis/pull/495

`A test is all we need.`

So please let us work this out - it's way too serious to keep posting instead of coding.

I'm not saying you shouldn't use templates, but be a little bit open and flexible in mind to additionally support ready-to-use PRs containing running examples.

``` java
         System.out.println(""); // violation
        System .out.println(""); // violation
        System. out.println(""); // violation
        System.out .println(""); // violation
        System.out. println(""); // violation
        System.out. println(""); // violation
        System.out.println (""); // violation
         "".equals(""); // violation
        "" .equals(""); // violation
        "". equals(""); // violation
        "".equals (""); // violation
        "".equals( ""); // violation
        "".equals("" ); // violation
        "".equals("") ; // violation
```
